# Record Matching

In [1]:
# importing the pandas library 
import pandas as pd

In [2]:
# getting the local file address
filePath = 'D:/Thapar/Placement/Organizations/NeenOpal/Records Matching Task.xlsx'

# storing the two sheets of data in different variables
data1 = pd.read_excel(filePath, sheet_name="data1")
data2 = pd.read_excel(filePath, sheet_name="data2")

In [3]:
# printing the first set of data
data1.head()

Order ID       Product ID  Qty
0  CA-2014-100006  TEC-PH-10002075    3
1  CA-2014-100090  FUR-TA-10003715    3
2  CA-2014-100090  OFF-BI-10001597    6
3  CA-2014-100293  OFF-PA-10000176    6
4  CA-2014-100328  OFF-BI-10000343    1

In [4]:
# printing the second set of data
data2.head()

Order ID       Product ID  Qty
0  CA-2014-100090  FUR-TA-10003715    3
1  CA-2014-100090  OFF-BI-10001597    6
2  CA-2014-100293  OFF-PA-10000176    6
3  CA-2014-100328  OFF-BI-10000343    1
4  CA-2014-100363  OFF-FA-10000611    2

In [5]:
# merging the columns to get another column of primary keys in each dataset
data1['Order_Product_1'] = data1['Order ID']+data1['Product ID']
data2['Order_Product_2'] = data2['Order ID']+data2['Product ID']

In [6]:
# printing the first modified set of data
data1.head()

Order ID       Product ID  Qty                Order_Product_1
0  CA-2014-100006  TEC-PH-10002075    3  CA-2014-100006TEC-PH-10002075
1  CA-2014-100090  FUR-TA-10003715    3  CA-2014-100090FUR-TA-10003715
2  CA-2014-100090  OFF-BI-10001597    6  CA-2014-100090OFF-BI-10001597
3  CA-2014-100293  OFF-PA-10000176    6  CA-2014-100293OFF-PA-10000176
4  CA-2014-100328  OFF-BI-10000343    1  CA-2014-100328OFF-BI-10000343

In [7]:
# printing the second modified set of data
data2.head()

Order ID       Product ID  Qty                Order_Product_2
0  CA-2014-100090  FUR-TA-10003715    3  CA-2014-100090FUR-TA-10003715
1  CA-2014-100090  OFF-BI-10001597    6  CA-2014-100090OFF-BI-10001597
2  CA-2014-100293  OFF-PA-10000176    6  CA-2014-100293OFF-PA-10000176
3  CA-2014-100328  OFF-BI-10000343    1  CA-2014-100328OFF-BI-10000343
4  CA-2014-100363  OFF-FA-10000611    2  CA-2014-100363OFF-FA-10000611

In [8]:
# performing the merge/join operation 
data = data1.merge(data2, how='outer', left_on=['Order_Product_1'], right_on=['Order_Product_2'])

#printing out the new dataset
data.head()

Order ID_x     Product ID_x  Qty_x                Order_Product_1  \
0  CA-2014-100006  TEC-PH-10002075    3.0  CA-2014-100006TEC-PH-10002075   
1  CA-2014-100090  FUR-TA-10003715    3.0  CA-2014-100090FUR-TA-10003715   
2  CA-2014-100090  OFF-BI-10001597    6.0  CA-2014-100090OFF-BI-10001597   
3  CA-2014-100293  OFF-PA-10000176    6.0  CA-2014-100293OFF-PA-10000176   
4  CA-2014-100328  OFF-BI-10000343    1.0  CA-2014-100328OFF-BI-10000343   

       Order ID_y     Product ID_y  Qty_y                Order_Product_2  
0             NaN              NaN    NaN                            NaN  
1  CA-2014-100090  FUR-TA-10003715    3.0  CA-2014-100090FUR-TA-10003715  
2  CA-2014-100090  OFF-BI-10001597    6.0  CA-2014-100090OFF-BI-10001597  
3  CA-2014-100293  OFF-PA-10000176    6.0  CA-2014-100293OFF-PA-10000176  
4  CA-2014-100328  OFF-BI-10000343    1.0  CA-2014-100328OFF-BI-10000343

In [9]:
# getting the shape of data1
data1.shape

(9478, 4)

In [10]:
# getting the shape of data2
data2.shape

(9479, 4)

# 1. How to identify the Records (Order ID + Product ID combination) present in data1 but missing in data2 (Specify the number of records missing in your answer)

In [11]:
# for finding the missing values we will check the columns where data2 is null
task1 = data[data['Order_Product_2'].isna()]
task1

Order ID_x     Product ID_x  Qty_x                Order_Product_1  \
0     CA-2014-100006  TEC-PH-10002075    3.0  CA-2014-100006TEC-PH-10002075   
10    CA-2014-100678  OFF-EN-10000056    3.0  CA-2014-100678OFF-EN-10000056   
19    CA-2014-100895  OFF-AR-10004511    2.0  CA-2014-100895OFF-AR-10004511   
35    CA-2014-101560  OFF-BI-10000309    3.0  CA-2014-101560OFF-BI-10000309   
61    CA-2014-102673  OFF-LA-10001771   12.0  CA-2014-102673OFF-LA-10001771   
...              ...              ...    ...                            ...   
9390  US-2017-160836  OFF-AP-10001626    2.0  US-2017-160836OFF-AP-10001626   
9403  US-2017-162558  FUR-FU-10002364    2.0  US-2017-162558FUR-FU-10002364   
9420  US-2017-163657  OFF-BI-10000138    5.0  US-2017-163657OFF-BI-10000138   
9427  US-2017-164056  FUR-TA-10001307    5.0  US-2017-164056FUR-TA-10001307   
9435  US-2017-165456  FUR-CH-10003981    6.0  US-2017-165456FUR-CH-10003981   

     Order ID_y Product ID_y  Qty_y Order_Product_2  
0           NaN          NaN    NaN             NaN  
10          NaN          NaN    NaN             NaN  
19          NaN          NaN    NaN             NaN  
35          NaN          NaN    NaN             NaN  
61          NaN          NaN    NaN             NaN  
...         ...          ...    ...             ...  
9390        NaN          NaN    NaN             NaN  
9403        NaN          NaN    NaN             NaN  
9420        NaN          NaN    NaN             NaN  
9427        NaN          NaN    NaN             NaN  
9435        NaN          NaN    NaN             NaN  

[507 rows x 8 columns]

Ans: 507

# 2. How to identify the Records (Order ID + Product ID combination) missing in data1 but present in data2 (Specify the number of records missing in your answer)

In [12]:
# for finding the missing values we will check the columns where data1 is null
task2 = data[data['Order_Product_1'].isna()]
task2

Order ID_x Product ID_x  Qty_x Order_Product_1      Order ID_y  \
9478        NaN          NaN    NaN             NaN  CA-2014-100706   
9479        NaN          NaN    NaN             NaN  CA-2014-100762   
9480        NaN          NaN    NaN             NaN  CA-2014-101427   
9481        NaN          NaN    NaN             NaN  CA-2014-102652   
9482        NaN          NaN    NaN             NaN  CA-2014-102869   
...         ...          ...    ...             ...             ...   
9981        NaN          NaN    NaN             NaN  US-2017-165344   
9982        NaN          NaN    NaN             NaN  US-2017-165358   
9983        NaN          NaN    NaN             NaN  US-2017-167920   
9984        NaN          NaN    NaN             NaN  US-2017-169502   
9985        NaN          NaN    NaN             NaN  US-2017-169551   

         Product ID_y  Qty_y                Order_Product_2  
9478  TEC-AC-10001314    2.0  CA-2014-100706TEC-AC-10001314  
9479  OFF-PA-10001815    3.0  CA-2014-100762OFF-PA-10001815  
9480  OFF-AR-10002257    3.0  CA-2014-101427OFF-AR-10002257  
9481  FUR-FU-10001918    7.0  CA-2014-102652FUR-FU-10001918  
9482  OFF-PA-10000788    3.0  CA-2014-102869OFF-PA-10000788  
...               ...    ...                            ...  
9981  OFF-BI-10003196   10.0  US-2017-165344OFF-BI-10003196  
9982  TEC-CO-10001943    5.0  US-2017-165358TEC-CO-10001943  
9983  OFF-AP-10000159    5.0  US-2017-167920OFF-AP-10000159  
9984  OFF-AP-10001947    5.0  US-2017-169502OFF-AP-10001947  
9985  FUR-BO-10001519    3.0  US-2017-169551FUR-BO-10001519  

[508 rows x 8 columns]

Ans: 508

# 3. Find the Sum of the total Qty of Records missing in data1 but present in data2

In [13]:
# using the output of task2 where we identified the Records (Order ID + Product ID combination) missing in data1 but
# present in data2 (Specify the number of records missing in your answer)
task3 = task2['Qty_y'].sum()
task3

1956.0

Ans: 1956

# 4. Find the total number of unique records (Order ID + Product ID combination) present in the combined dataset of data1 and data2

In [14]:
# the total combinations can be found through the combination found earlier and stored in data
task4 = data.shape[0]
task4

9986

Ans: 9986

# Q2. Python 

In [18]:
# installing the mysql library to link python with mysql database
!pip install mysql-connector-python

# 1. Read the csv or excel file.

In [19]:
# reading into the dataframe the bikes data
dataframe = pd.read_csv('D:/Thapar/Placement/Organizations/NeenOpal/used_bikes.csv')

# printing the dataset
dataframe.head()

bike_name   price       city  kms_driven  \
0    TVS Star City Plus Dual Tone 110cc   35000  Ahmedabad       17654   
1           Royal Enfield Classic 350cc  119900      Delhi       11000   
2                  Triumph Daytona 675R  600000      Delhi         110   
3                  TVS Apache RTR 180cc   65000  Bangalore       16329   
4  Yamaha FZ S V 2.0 150cc-Ltd. Edition   80000  Bangalore       10000   

         owner  age  power          brand  
0  First Owner    3    110            TVS  
1  First Owner    4    350  Royal Enfield  
2  First Owner    8    675        Triumph  
3  First Owner    4    180            TVS  
4  First Owner    3    150         Yamaha

In [20]:
# getting the dataframe types
dataframe.dtypes

bike_name     object
price          int64
city          object
kms_driven     int64
owner         object
age            int64
power          int64
brand         object
dtype: object

In [21]:
# finding the individual memory usage before modifying the datatypes
dataframe.memory_usage()

Index          128
bike_name     1192
price         1192
city          1192
kms_driven    1192
owner         1192
age           1192
power         1192
brand         1192
dtype: int64

In [22]:
# checking the memory usage before the data type changes
initialMemoryUse = dataframe.memory_usage().sum()
initialMemoryUse

9664

# Change the datatypes of the column and check the memory usage before and after the change in the data types.

In [23]:
# changing the data types of the float columns to int32
float_columns = ['price', 'kms_driven', 'age', 'power']
for column in float_columns:
    dataframe[column] = dataframe[column].astype('int32')

In [24]:
# finding the individual memory usage after modifying the datatypes
dataframe.memory_usage()

Index          128
bike_name     1192
price          596
city          1192
kms_driven     596
owner         1192
age            596
power          596
brand         1192
dtype: int64

In [25]:
# checking the total memory usage before the data type changes
finalMemoryUse = dataframe.memory_usage().sum()
finalMemoryUse

7280

In [26]:
# removing indexing from the table
dataframe.reset_index(drop=True, inplace=True)
print(dataframe.head())

                              bike_name   price       city  kms_driven  \
0    TVS Star City Plus Dual Tone 110cc   35000  Ahmedabad       17654   
1           Royal Enfield Classic 350cc  119900      Delhi       11000   
2                  Triumph Daytona 675R  600000      Delhi         110   
3                  TVS Apache RTR 180cc   65000  Bangalore       16329   
4  Yamaha FZ S V 2.0 150cc-Ltd. Edition   80000  Bangalore       10000   

         owner  age  power          brand  
0  First Owner    3    110            TVS  
1  First Owner    4    350  Royal Enfield  
2  First Owner    8    675        Triumph  
3  First Owner    4    180            TVS  
4  First Owner    3    150         Yamaha  


In [27]:
# importing the libraries for MySQL
import mysql.connector

In [28]:
# preparing the linkage of the database connectors with MySQL
dataBase_host = "localhost"
dataBase_user = "root"
dataBase_password = "Punam&mom2"
dataBase_name = "bikes"
table_name = "used_bikes" 

# connecting to the MySQL database
connection = mysql.connector.connect(host=dataBase_host, user=dataBase_user, password=dataBase_password, database=dataBase_name)
cursor = connection.cursor()

In [29]:
# dropping the index from the exsisting table
try:
    drop_index_query = f"DROP INDEX IF EXISTS idx_{table_name}_bike_name ON {table_name};"
    cursor.execute(drop_index_query)
    connection.commit()
except mysql.connector.Error as e:
    pass

In [30]:
# this is not a necessary step. it is done here as initially there is not data in the table
# defining the table creation query
create_table_query = """
CREATE TABLE IF NOT EXISTS {} (
    bike_name VARCHAR(255),
    price INT,
    city VARCHAR(255),
    kms_driven INT,
    owner VARCHAR(255),
    age INT,
    power INT,
    brand VARCHAR(255)
);
""".format(table_name)

# executing the table creation query
cursor.execute(create_table_query)

# commiting the changes to the database
connection.commit()

# Dump the data into the mysql database.

In [31]:
# inserting data into the table
for index, row in dataframe.iterrows():
    values = tuple(row)
    insert_query = f"INSERT INTO {table_name} (bike_name, price, city, kms_driven, owner, age, power, brand) VALUES (%s, %s, %s, %s, %s, %s, %s, %s);"
    cursor.execute(insert_query, values)

In [32]:
# commiting the changes to the database
connection.commit()

# index the column after the data is inserted.

In [33]:
# adding a new auto-incremented primary key column 'id'
add_id_query = "ALTER TABLE {} ADD id INT AUTO_INCREMENT PRIMARY KEY;".format(table_name)
cursor.execute(add_id_query)
connection.commit()

# index the 'id' column
index_query = f"CREATE INDEX idx_{table_name}_id ON {table_name}(id);"
cursor.execute(index_query)
connection.commit()

In [34]:
# commiting the changes to the database
connection.commit()

In [35]:
# showing the indexed data
cursor.execute(f"SELECT * FROM {table_name};")
rows = cursor.fetchall()
for row in rows:
    print(row)

('TVS Star City Plus Dual Tone 110cc', 35000, 'Ahmedabad', 17654, 'First Owner', 3, 110, 'TVS', 1)
('Royal Enfield Classic 350cc', 119900, 'Delhi', 11000, 'First Owner', 4, 350, 'Royal Enfield', 2)
('Triumph Daytona 675R', 600000, 'Delhi', 110, 'First Owner', 8, 675, 'Triumph', 3)
('TVS Apache RTR 180cc', 65000, 'Bangalore', 16329, 'First Owner', 4, 180, 'TVS', 4)
('Yamaha FZ S V 2.0 150cc-Ltd. Edition', 80000, 'Bangalore', 10000, 'First Owner', 3, 150, 'Yamaha', 5)
('Yamaha FZs 150cc', 53499, 'Delhi', 25000, 'First Owner', 6, 150, 'Yamaha', 6)
('Honda CB Hornet 160R  ABS DLX', 85000, 'Delhi', 8200, 'First Owner', 3, 160, 'Honda', 7)
('Hero Splendor Plus Self Alloy 100cc', 45000, 'Delhi', 12645, 'First Owner', 3, 100, 'Hero', 8)
('Royal Enfield Thunderbird X 350cc', 145000, 'Bangalore', 9190, 'First Owner', 3, 350, 'Royal Enfield', 9)
('Royal Enfield Classic Desert Storm 500cc', 88000, 'Delhi', 19000, 'Second Owner', 7, 500, 'Royal Enfield', 10)
('Yamaha YZF-R15 2.0 150cc', 72000, 'Ban